src/extension/gogoduck/src/main/java/au/org/emii/gogoduck/worker/GoGoDuck.java

```diff
@@ -0,0 +1,511 @@
+package au.org.emii.gogoduck.worker;
+
+import java.io.*;
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.net.URL;
+import java.nio.channels.Channels;
+import java.nio.channels.ReadableByteChannel;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+
+import org.apache.commons.io.FileUtils;
+import org.apache.commons.io.FilenameUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.opengis.util.ProgressListener;
+
+import java.io.IOException;
+import java.util.*;
+import java.util.zip.GZIPInputStream;
+
+public class GoGoDuck {
+    private static final Logger logger = LoggerFactory.getLogger(GoGoDuck.class);
+
+    private static final Map<String, String> replacePrefixes = new HashMap<String, String>();
+    static {
+        replacePrefixes.put("/mnt/imos-t3/", "https://data.aodn.org.au/");
+        replacePrefixes.put("/mnt/opendap/2/SRS/", "https://thredds.aodn.org.au/thredds/fileServer/srs/");
+    }
+
+    public static final String ncksPath = "/usr/bin/ncks";
+    public static final String ncrcatPath = "/usr/bin/ncrcat";
+    public static final String ncdpqPath = "/usr/bin/ncpdq";
+
+    private final String geoserver;
+    private final String profile;
+    private final String subset;
+    private final Path outputFile;
+    private final Integer limit;
+    private Path baseTmpDir;
+    private int threadCount = 1;
+    private UserLog userLog = null;
+    private ProgressListener progressListener = null;
+
+    public GoGoDuck(String geoserver, String profile, String subset, String outputFile, Integer limit) {
+        this.geoserver = geoserver;
+        this.profile = profile;
+        this.subset = subset;
+        this.outputFile = new File(outputFile).toPath();
+        this.limit = limit;
+        this.baseTmpDir = new File(System.getProperty("java.io.tmpdir")).toPath();
+        this.userLog = new UserLog();
+    }
+
+    public void setTmpDir(String tmpDir) {
```

Comment:
Any reason to have the settable?  Or can we just make them part of the constructor, and immutable?